// moneroo/dist/cjs/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Moneroo = void 0;
const errors_js_1 = require("./errors.js");
const payments_js_1 = require("./resources/payments.js");
const payouts_js_1 = require("./resources/payouts.js");
const webhooks_js_1 = require("./resources/webhooks.js");
const DEFAULT_BASE_URL = 'https://api.moneroo.io/v1';
class Moneroo {
    constructor(options) {
        if (!options.secretKey) {
            throw new Error('secretKey is required');
        }
        this.secretKey = options.secretKey;
        this.baseUrl = (options.baseUrl ?? DEFAULT_BASE_URL).replace(/\/$/, '');
        this.payments = new payments_js_1.PaymentsResource(this);
        this.payouts = new payouts_js_1.PayoutsResource(this);
        this.webhooks = new webhooks_js_1.WebhooksResource(options.webhookSecret);
    }
    async _request(method, path, body) {
        const url = `${this.baseUrl}/${path.replace(/^\//, '')}`;
        const headers = {
            Authorization: `Bearer ${this.secretKey}`,
            Accept: 'application/json',
            'Content-Type': 'application/json',
        };
        const response = await fetch(url, {
            method,
            headers,
            body: body !== undefined ? JSON.stringify(body) : undefined,
        });
        let payload;
        try {
            payload = (await response.json());
        }
        catch {
            throw new errors_js_1.MonerooError(`Failed to parse response (HTTP ${response.status})`, response.status);
        }
        if (response.ok) {
            return payload;
        }
        const { message, errors } = payload;
        switch (response.status) {
            case 401:
                throw new errors_js_1.AuthenticationError(message, 401, errors);
            case 403:
                throw new errors_js_1.AuthenticationError(message, 403, errors);
            case 404:
                throw new errors_js_1.NotFoundError(message, errors);
            case 400:
                throw new errors_js_1.ValidationError(message, 400, errors);
            case 422:
                throw new errors_js_1.ValidationError(message, 422, errors);
            case 429:
                throw new errors_js_1.RateLimitError(message, errors);
            case 500:
                throw new errors_js_1.ServerError(message, 500, errors);
            case 503:
                throw new errors_js_1.ServerError(message, 503, errors);
            default:
                throw new errors_js_1.MonerooError(message, response.status, errors);
        }
    }
    /** @internal Exposed for resource classes to call */
    _get(path) {
        return this._request('GET', path);
    }
    _post(path, body) {
        return this._request('POST', path, body);
    }
    _put(path, body) {
        return this._request('PUT', path, body);
    }
    _patch(path, body) {
        return this._request('PATCH', path, body);
    }
    _delete(path) {
        return this._request('DELETE', path);
    }
}
exports.Moneroo = Moneroo;
