// moneroo/dist/cjs/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerError = exports.RateLimitError = exports.NotFoundError = exports.ValidationError = exports.AuthenticationError = exports.MonerooError = void 0;
class MonerooError extends Error {
    constructor(message, status, errors = null) {
        super(message);
        this.name = 'MonerooError';
        this.status = status;
        this.errors = errors;
        // Restore prototype chain for instanceof checks
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
exports.MonerooError = MonerooError;
/** 401 - Missing or invalid API key / 403 - Insufficient access rights */
class AuthenticationError extends MonerooError {
    constructor(message, status, errors = null) {
        super(message, status, errors);
        this.name = 'AuthenticationError';
    }
}
exports.AuthenticationError = AuthenticationError;
/** 400 - Malformed request or missing parameters / 422 - Parameters invalid for the request */
class ValidationError extends MonerooError {
    constructor(message, status, errors = null) {
        super(message, status, errors);
        this.name = 'ValidationError';
    }
}
exports.ValidationError = ValidationError;
/** 404 - Resource does not exist */
class NotFoundError extends MonerooError {
    constructor(message, errors = null) {
        super(message, 404, errors);
        this.name = 'NotFoundError';
    }
}
exports.NotFoundError = NotFoundError;
/**
 * 429 - Rate limit exceeded (120 req/min).
 * Wait 60 seconds before retrying.
 */
class RateLimitError extends MonerooError {
    constructor(message, errors = null) {
        super(message, 429, errors);
        /** Recommended wait time in milliseconds before retrying */
        this.retryAfterMs = 60000;
        this.name = 'RateLimitError';
    }
}
exports.RateLimitError = RateLimitError;
/** 500 / 503 - Server-side error */
class ServerError extends MonerooError {
    constructor(message, status, errors = null) {
        super(message, status, errors);
        this.name = 'ServerError';
    }
}
exports.ServerError = ServerError;
