// moneroo/dist/esm/errors.js
export class MonerooError extends Error {
    constructor(message, status, errors = null) {
        super(message);
        this.name = 'MonerooError';
        this.status = status;
        this.errors = errors;
        // Restore prototype chain for instanceof checks
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
/** 401 - Missing or invalid API key / 403 - Insufficient access rights */
export class AuthenticationError extends MonerooError {
    constructor(message, status, errors = null) {
        super(message, status, errors);
        this.name = 'AuthenticationError';
    }
}
/** 400 - Malformed request or missing parameters / 422 - Parameters invalid for the request */
export class ValidationError extends MonerooError {
    constructor(message, status, errors = null) {
        super(message, status, errors);
        this.name = 'ValidationError';
    }
}
/** 404 - Resource does not exist */
export class NotFoundError extends MonerooError {
    constructor(message, errors = null) {
        super(message, 404, errors);
        this.name = 'NotFoundError';
    }
}
/**
 * 429 - Rate limit exceeded (120 req/min).
 * Wait 60 seconds before retrying.
 */
export class RateLimitError extends MonerooError {
    constructor(message, errors = null) {
        super(message, 429, errors);
        /** Recommended wait time in milliseconds before retrying */
        this.retryAfterMs = 60000;
        this.name = 'RateLimitError';
    }
}
/** 500 / 503 - Server-side error */
export class ServerError extends MonerooError {
    constructor(message, status, errors = null) {
        super(message, status, errors);
        this.name = 'ServerError';
    }
}
