// cinetpay-js/dist/index.js
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined") return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), member.set(obj, value), value);

// src/types/config.ts
var DEFAULT_BASE_URL = "https://api.cinetpay.net";
var PRODUCTION_BASE_URL = "https://api.cinetpay.co";
var API_KEY_PREFIX_TEST = "sk_test_";
var API_KEY_PREFIX_LIVE = "sk_live_";
var DEFAULT_TOKEN_TTL = 82800;
var DEFAULT_TIMEOUT = 3e4;

// src/errors/cinetpay-error.ts
var CinetPayError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "CinetPayError";
  }
};

// src/errors/api-error.ts
var ApiError = class _ApiError extends CinetPayError {
  constructor(apiCode, apiStatus, description) {
    super(`[${apiCode}] ${apiStatus}: ${description}`);
    this.name = "ApiError";
    this.apiCode = apiCode;
    this.apiStatus = apiStatus;
    this.description = description;
  }
  /**
   * Construit une ApiError à partir d'une réponse brute de l'API.
   * @internal
   */
  static fromResponse(data) {
    return new _ApiError(
      data.code ?? 0,
      data.status ?? "UNKNOWN",
      data.description ?? data.message ?? ""
    );
  }
};

// src/errors/authentication-error.ts
var AuthenticationError = class extends CinetPayError {
  constructor(message = "Authentication failed") {
    super(message);
    this.name = "AuthenticationError";
  }
};

// src/errors/index.ts
var NetworkError = class extends CinetPayError {
  constructor(message, cause) {
    super(message);
    this.name = "NetworkError";
    this.cause = cause;
  }
};
var TimeoutError = class extends CinetPayError {
  constructor(timeoutMs) {
    super(`Request timed out after ${timeoutMs}ms`);
    this.name = "TimeoutError";
  }
};

// src/http/http-client.ts
var HttpClient = class {
  constructor(baseUrl, timeout, logger) {
    this.baseUrl = baseUrl;
    this.timeout = timeout;
    this.logger = logger;
    if (typeof globalThis.fetch === "undefined") {
      throw new Error(
        "fetch is not available in this environment. Use Node.js 18+ or provide a fetch polyfill."
      );
    }
  }
  /** Effectue une requête POST et retourne la réponse parsée */
  async post(path, body, token) {
    return this.request("POST", path, body, token);
  }
  /** Effectue une requête GET et retourne la réponse parsée */
  async get(path, token) {
    return this.request("GET", path, void 0, token);
  }
  async request(method, path, body, token) {
    const url = `${this.baseUrl}${path}`;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);
    const headers = {
      Accept: "application/json",
      "Content-Type": "application/json"
    };
    if (token) {
      headers.Authorization = `Bearer ${token}`;
    }
    this.logger.debug(`${method} ${path}`, {
      ...body ? { body: this.sanitizeBody(body) } : {}
    });
    try {
      const response = await globalThis.fetch(url, {
        method,
        headers,
        body: body ? JSON.stringify(body) : void 0,
        signal: controller.signal
      });
      const data = await response.json();
      this.logger.debug(`${method} ${path} \u2192 ${response.status}`, {
        code: data.code,
        status: data.status
      });
      return this.handleResponse(data, response.status, method, path);
    } catch (error) {
      if (error instanceof ApiError) throw error;
      if (error instanceof DOMException && error.name === "AbortError") {
        this.logger.error(`${method} ${path} \u2192 timeout (${this.timeout}ms)`);
        throw new TimeoutError(this.timeout);
      }
      this.logger.error(`${method} ${path} \u2192 network error`, {
        message: error instanceof Error ? error.message : String(error)
      });
      throw new NetworkError(
        `Request to ${method} ${path} failed`,
        error
      );
    } finally {
      clearTimeout(timeoutId);
    }
  }
  handleResponse(data, httpStatus, method, path) {
    const code = data.code;
    if (httpStatus >= 400 || code !== void 0 && code !== 200 && code !== 100 && code !== 2001 && code !== 2002) {
      if (data.description || code !== void 0 && httpStatus >= 400) {
        const error = ApiError.fromResponse(data);
        this.logger.error(`${method} ${path} \u2192 API error`, {
          apiCode: error.apiCode,
          apiStatus: error.apiStatus,
          description: error.description
        });
        throw error;
      }
    }
    return data;
  }
  /** Masque les champs sensibles (credentials, tokens) dans les logs */
  sanitizeBody(body) {
    if (typeof body !== "object" || body === null) return body;
    const sanitized = { ...body };
    for (const key of Object.keys(sanitized)) {
      const lower = key.toLowerCase();
      if (lower.includes("password") || lower.includes("secret") || lower.includes("token") || lower.includes("key") && lower.includes("api")) {
        sanitized[key] = "***";
      }
    }
    return sanitized;
  }
};

// src/auth/authenticator.ts
var _apiKey, _apiPassword, _pendingAuth;
var Authenticator = class {
  /**
   * @param httpClient - Client HTTP partagé entre toutes les APIs
   * @param country - Code pays ISO en majuscules (ex: 'CI', 'SN')
   * @param credentials - Couple `apiKey` / `apiPassword` pour ce pays
   * @param tokenTtl - Durée de cache du token en secondes (défaut: 82800 = 23h)
   * @param tokenStore - Backend de cache (MemoryTokenStore par défaut)
   * @param logger - Logger pour le debug et les avertissements
   */
  constructor(httpClient, country, credentials, tokenTtl, tokenStore, logger) {
    /** Credentials stockés en champs privés ES2022 — invisibles à la sérialisation */
    __privateAdd(this, _apiKey);
    __privateAdd(this, _apiPassword);
    /** Promise en cours pour éviter les appels auth simultanés (token stampede) */
    __privateAdd(this, _pendingAuth, null);
    this.httpClient = httpClient;
    this.country = country;
    this.tokenTtl = tokenTtl;
    this.tokenStore = tokenStore;
    this.logger = logger;
    this.cacheKey = `cinetpay_token_${country.toLowerCase()}`;
    __privateSet(this, _apiKey, credentials.apiKey);
    __privateSet(this, _apiPassword, credentials.apiPassword);
  }
  /**
   * Récupère un token JWT valide.
   * Retourne le token en cache s'il est encore valide,
   * sinon appelle l'API d'authentification.
   *
   * Utilise le promise coalescing pour éviter les appels simultanés
   * (token stampede) : si N requêtes arrivent en même temps avec un cache vide,
   * un seul appel auth est effectué et les N requêtes partagent la même promise.
   *
   * @returns Token JWT Bearer
   * @throws {AuthenticationError} Si l'API ne retourne pas de token
   */
  async getToken() {
    const cached = await this.tokenStore.get(this.cacheKey);
    if (cached) {
      this.logger.debug(`Token cache hit for ${this.country}`);
      return cached;
    }
    this.logger.debug(`Token cache miss for ${this.country}, authenticating...`);
    if (!__privateGet(this, _pendingAuth)) {
      __privateSet(this, _pendingAuth, this.authenticate().finally(() => {
        __privateSet(this, _pendingAuth, null);
      }));
    } else {
      this.logger.debug(`Reusing pending auth for ${this.country} (coalesced)`);
    }
    return __privateGet(this, _pendingAuth);
  }
  /**
   * Force le renouvellement du token (supprime le cache puis ré-authentifie).
   * Appelé automatiquement par les APIs quand l'API retourne EXPIRED_TOKEN (1003).
   *
   * @returns Nouveau token JWT
   * @throws {AuthenticationError} Si l'API ne retourne pas de token
   */
  async forceRefresh() {
    this.logger.warn(`Force refreshing token for ${this.country} (expired or invalid)`);
    await this.tokenStore.delete(this.cacheKey);
    return this.authenticate();
  }
  /**
   * Supprime le token en cache pour ce pays.
   * Utile pour forcer une ré-authentification au prochain appel.
   */
  async clearCache() {
    await this.tokenStore.delete(this.cacheKey);
  }
  /**
   * Empêche la fuite des credentials dans `JSON.stringify()`.
   * Retourne uniquement le pays et la clé de cache.
   */
  toJSON() {
    return { country: this.country, cacheKey: this.cacheKey };
  }
  /**
   * Appelle `POST /v1/oauth/login` avec les credentials du pays
   * et stocke le token reçu dans le cache.
   *
   * @returns Token JWT
   * @throws {AuthenticationError} Si la réponse ne contient pas `access_token`
   */
  async authenticate() {
    let data;
    try {
      data = await this.httpClient.post("/v1/oauth/login", {
        api_key: __privateGet(this, _apiKey),
        api_password: __privateGet(this, _apiPassword)
      });
    } catch (error) {
      throw new AuthenticationError(
        `Authentication failed for country ${this.country}`
      );
    }
    if (!data.access_token) {
      throw new AuthenticationError(
        `Authentication failed for country ${this.country}: no access_token in response`
      );
    }
    await this.tokenStore.set(this.cacheKey, data.access_token, this.tokenTtl);
    this.logger.debug(`Token cached for ${this.country} (TTL: ${this.tokenTtl}s)`);
    return data.access_token;
  }
};
_apiKey = new WeakMap();
_apiPassword = new WeakMap();
_pendingAuth = new WeakMap();

// src/auth/token-store.ts
var MemoryTokenStore = class {
  constructor() {
    this.cache = /* @__PURE__ */ new Map();
  }
  /**
   * Récupère un token depuis le cache.
   * Si le token est expiré, il est supprimé et `null` est retourné.
   *
   * @param key - Clé de cache (ex: `cinetpay_token_ci`)
   * @returns Le token ou `null` s'il est absent/expiré
   */
  get(key) {
    const entry = this.cache.get(key);
    if (!entry) return null;
    if (Date.now() >= entry.expiresAt) {
      this.cache.delete(key);
      return null;
    }
    return entry.value;
  }
  /**
   * Stocke un token avec une durée de vie.
   *
   * @param key - Clé de cache
   * @param value - Token JWT
   * @param ttlSeconds - Durée de vie en secondes
   */
  set(key, value, ttlSeconds) {
    this.cache.set(key, {
      value,
      expiresAt: Date.now() + ttlSeconds * 1e3
    });
  }
  /**
   * Supprime un token du cache.
   *
   * @param key - Clé de cache à supprimer
   */
  delete(key) {
    this.cache.delete(key);
  }
};

// src/types/payment.ts
function serializePaymentRequest(req) {
  const body = {
    currency: req.currency,
    merchant_transaction_id: req.merchantTransactionId,
    amount: req.amount,
    lang: req.lang,
    designation: req.designation,
    client_email: req.clientEmail,
    client_first_name: req.clientFirstName,
    client_last_name: req.clientLastName,
    success_url: req.successUrl,
    failed_url: req.failedUrl,
    notify_url: req.notifyUrl,
    channel: req.channel
  };
  if (req.paymentMethod !== void 0) body.payment_method = req.paymentMethod;
  if (req.clientPhoneNumber !== void 0)
    body.client_phone_number = req.clientPhoneNumber;
  if (req.directPay !== void 0) body.direct_pay = req.directPay;
  if (req.otpCode !== void 0) body.otp_code = req.otpCode;
  return body;
}
function toPaymentResponse(raw) {
  const details = raw.details;
  return {
    code: raw.code,
    status: raw.status,
    paymentToken: raw.payment_token,
    notifyToken: raw.notify_token,
    transactionId: raw.transaction_id,
    merchantTransactionId: raw.merchant_transaction_id,
    paymentUrl: raw.payment_url,
    details: details ? {
      code: details.code,
      status: details.status,
      message: details.message,
      mustBeRedirected: details.must_be_redirected
    } : { code: 0, status: "INITIATED", message: "", mustBeRedirected: false }
  };
}
function toPaymentStatus(raw) {
  const user = raw.user;
  return {
    code: raw.code,
    status: raw.status,
    merchantTransactionId: raw.merchant_transaction_id,
    transactionId: raw.transaction_id,
    user: user ? {
      name: user.name,
      email: user.email,
      phoneNumber: user.phone_number
    } : { name: "", email: "", phoneNumber: "" }
  };
}

// src/constants/currencies.ts
var CURRENCIES = {
  /** Franc CFA BCEAO (Afrique de l'Ouest) */
  XOF: "XOF",
  /** Franc CFA BEAC (Afrique Centrale) */
  XAF: "XAF",
  /** Franc guinéen */
  GNF: "GNF",
  /** Franc congolais */
  CDF: "CDF",
  /** Dollar américain */
  USD: "USD"
};

// src/constants/channels.ts
var CHANNELS = {
  /** Notification push envoyée au téléphone du client */
  PUSH: "PUSH",
  /** Code OTP saisi par le client */
  OTP: "OTP",
  /** QR Code scanné par le client */
  QRCODE: "QRCODE"
};

// src/constants/payment-methods.ts
var PAYMENT_METHODS = {
  // Côte d'Ivoire
  /** Orange Money — Côte d'Ivoire */
  OM_CI: "OM_CI",
  /** Moov Money — Côte d'Ivoire */
  MOOV_CI: "MOOV_CI",
  /** MTN Mobile Money — Côte d'Ivoire */
  MTN_CI: "MTN_CI",
  /** Wave — Côte d'Ivoire */
  WAVE_CI: "WAVE_CI",
  // Burkina Faso
  /** Orange Money — Burkina Faso */
  OM_BF: "OM_BF",
  /** Moov Money — Burkina Faso */
  MOOV_BF: "MOOV_BF",
  /** Wave — Burkina Faso */
  WAVE_BF: "WAVE_BF",
  // Mali
  /** Orange Money — Mali */
  OM_ML: "OM_ML",
  /** Moov Money — Mali */
  MOOV_ML: "MOOV_ML",
  // Sénégal
  /** Orange Money — Sénégal */
  OM_SN: "OM_SN",
  /** Free Money — Sénégal */
  FREE_SN: "FREE_SN",
  /** Expresso — Sénégal */
  EXPRESSO_SN: "EXPRESSO_SN",
  /** Wave — Sénégal */
  WAVE_SN: "WAVE_SN",
  // Togo
  /** Moov Money — Togo */
  MOOV_TG: "MOOV_TG",
  /** TMoney (Togocell) — Togo */
  TMONEY_TG: "TMONEY_TG",
  // Guinée
  /** Orange Money — Guinée */
  OM_GN: "OM_GN",
  /** MTN Mobile Money — Guinée */
  MTN_GN: "MTN_GN",
  // Cameroun
  /** Orange Money — Cameroun */
  OM_CM: "OM_CM",
  /** MTN Mobile Money — Cameroun */
  MTN_CM: "MTN_CM",
  // Bénin
  /** Moov Money — Bénin */
  MOOV_BJ: "MOOV_BJ",
  /** MTN Mobile Money — Bénin */
  MTN_BJ: "MTN_BJ",
  // RD Congo
  /** Orange Money — RD Congo */
  OM_CD: "OM_CD",
  /** Airtel Money — RD Congo */
  AIRTEL_CD: "AIRTEL_CD",
  /** M-Pesa — RD Congo */
  MPESA_CD: "MPESA_CD",
  /** Africell — RD Congo */
  AFRICELL_CD: "AFRICELL_CD",
  // Niger
  /** Airtel Money — Niger */
  AIRTEL_NE: "AIRTEL_NE",
  /** Moov Money — Niger */
  MOOV_NE: "MOOV_NE",
  /** Zamani — Niger */
  ZAMANI_NE: "ZAMANI_NE"
};
var PAYMENT_METHODS_BY_COUNTRY = {
  CI: ["OM_CI", "MOOV_CI", "MTN_CI", "WAVE_CI"],
  BF: ["OM_BF", "MOOV_BF", "WAVE_BF"],
  ML: ["OM_ML", "MOOV_ML"],
  SN: ["OM_SN", "FREE_SN", "EXPRESSO_SN", "WAVE_SN"],
  TG: ["MOOV_TG", "TMONEY_TG"],
  GN: ["OM_GN", "MTN_GN"],
  CM: ["OM_CM", "MTN_CM"],
  BJ: ["MOOV_BJ", "MTN_BJ"],
  CD: ["OM_CD", "AIRTEL_CD", "MPESA_CD", "AFRICELL_CD"],
  NE: ["AIRTEL_NE", "MOOV_NE", "ZAMANI_NE"]
};

// src/validation.ts
var ValidationError = class extends TypeError {
  constructor(field, message) {
    super(`[${field}] ${message}`);
    this.name = "ValidationError";
  }
};
var CURRENCY_VALUES = new Set(Object.values(CURRENCIES));
var CHANNEL_VALUES = new Set(Object.values(CHANNELS));
var PAYMENT_METHOD_VALUES = new Set(Object.values(PAYMENT_METHODS));
var EMAIL_REGEX = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
var PHONE_REGEX = /^\+\d{8,15}$/;
var OTP_REGEX = /^\d{4,6}$/;
function assertString(field, value, min, max) {
  if (typeof value !== "string" || value.length < min || value.length > max) {
    throw new ValidationError(field, `must be a string between ${min} and ${max} characters (got ${typeof value === "string" ? value.length : typeof value})`);
  }
}
function assertEnum(field, value, allowed) {
  if (!allowed.has(value)) {
    throw new ValidationError(field, `must be one of: ${[...allowed].join(", ")} (got "${value}")`);
  }
}
function assertAmount(field, value, min, max) {
  if (typeof value !== "number" || !Number.isInteger(value) || value < min || value > max) {
    throw new ValidationError(field, `must be an integer between ${min} and ${max} (got ${value})`);
  }
}
function assertUrl(field, value, maxLength) {
  assertString(field, value, 1, maxLength);
  if (!value.startsWith("http://") && !value.startsWith("https://")) {
    throw new ValidationError(field, `must be a valid URL starting with http:// or https://`);
  }
}
function assertEmail(field, value) {
  if (typeof value !== "string" || !EMAIL_REGEX.test(value)) {
    throw new ValidationError(field, `must be a valid email address (got "${value}")`);
  }
}
function assertPhone(field, value) {
  if (!PHONE_REGEX.test(value)) {
    throw new ValidationError(field, `must be in international format +XXXXXXXXXXXX (got "${value}")`);
  }
}
function validatePaymentRequest(req) {
  assertEnum("currency", req.currency, CURRENCY_VALUES);
  assertString("merchantTransactionId", req.merchantTransactionId, 1, 30);
  assertAmount("amount", req.amount, 100, 25e5);
  assertEnum("lang", req.lang, /* @__PURE__ */ new Set(["fr", "en"]));
  assertString("designation", req.designation, 1, 255);
  assertEmail("clientEmail", req.clientEmail);
  assertString("clientFirstName", req.clientFirstName, 2, 255);
  assertString("clientLastName", req.clientLastName, 2, 255);
  assertUrl("successUrl", req.successUrl, 120);
  assertUrl("failedUrl", req.failedUrl, 120);
  assertUrl("notifyUrl", req.notifyUrl, 120);
  assertEnum("channel", req.channel, CHANNEL_VALUES);
  if (req.paymentMethod !== void 0) {
    assertEnum("paymentMethod", req.paymentMethod, PAYMENT_METHOD_VALUES);
  }
  if (req.clientPhoneNumber !== void 0) {
    assertPhone("clientPhoneNumber", req.clientPhoneNumber);
  }
  if (req.otpCode !== void 0) {
    if (typeof req.otpCode !== "string" || !OTP_REGEX.test(req.otpCode)) {
      throw new ValidationError("otpCode", "must be 4 to 6 digits");
    }
  }
  if (req.directPay === true) {
    if (!req.clientPhoneNumber) {
      throw new ValidationError("clientPhoneNumber", "is required when directPay is true");
    }
    if (!req.paymentMethod) {
      throw new ValidationError("paymentMethod", "is required when directPay is true");
    }
  }
}
function validateTransferRequest(req) {
  assertEnum("currency", req.currency, CURRENCY_VALUES);
  assertString("merchantTransactionId", req.merchantTransactionId, 1, 30);
  assertAmount("amount", req.amount, 100, 15e5);
  assertEnum("paymentMethod", req.paymentMethod, PAYMENT_METHOD_VALUES);
  assertString("reason", req.reason, 1, 255);
  assertUrl("notifyUrl", req.notifyUrl, 120);
  assertPhone("phoneNumber", req.phoneNumber);
}

// src/api/payment.ts
var PaymentApi = class {
  constructor(httpClient, authenticators) {
    this.httpClient = httpClient;
    this.authenticators = authenticators;
  }
  /**
   * Initialise une transaction de paiement web.
   *
   * @param request - Paramètres du paiement (montant, devise, URLs, client, etc.)
   * @param country - Code pays ISO (ex: 'CI', 'SN') — détermine les credentials utilisés
   * @returns Réponse contenant le `paymentUrl` (redirection) et le `paymentToken` (vérification)
   * @throws {ApiError} Si l'API retourne une erreur (ex: TRANSACTION_EXIST, INVALID_PARAMS)
   * @throws {TypeError} Si le pays n'est pas configuré dans les credentials
   *
   * @example
   * ```typescript
   * const payment = await client.payment.initialize({
   *   currency: 'XOF',
   *   merchantTransactionId: 'ORDER-001',
   *   amount: 1000,
   *   lang: 'fr',
   *   designation: 'Achat en ligne',
   *   clientEmail: 'client@email.com',
   *   clientFirstName: 'Jean',
   *   clientLastName: 'Dupont',
   *   successUrl: 'https://monsite.com/success',
   *   failedUrl: 'https://monsite.com/failed',
   *   notifyUrl: 'https://monsite.com/webhook',
   *   channel: 'PUSH',
   * }, 'CI')
   *
   * // Rediriger le client si nécessaire
   * if (payment.details.mustBeRedirected) {
   *   redirect(payment.paymentUrl)
   * }
   * ```
   */
  async initialize(request, country) {
    validatePaymentRequest(request);
    const auth = this.resolveAuthenticator(country);
    const body = serializePaymentRequest(request);
    try {
      const token = await auth.getToken();
      const raw = await this.httpClient.post(
        "/v1/payment",
        body,
        token
      );
      return toPaymentResponse(raw);
    } catch (error) {
      if (error instanceof ApiError && this.isTokenExpired(error)) {
        const token = await auth.forceRefresh();
        const raw = await this.httpClient.post(
          "/v1/payment",
          body,
          token
        );
        return toPaymentResponse(raw);
      }
      throw error;
    }
  }
  /**
   * Récupère le statut actuel d'un paiement.
   *
   * @param identifier - `payment_token`, `transaction_id` ou `merchant_transaction_id`
   * @param country - Code pays ISO (ex: 'CI', 'SN')
   * @returns Statut de la transaction avec les informations utilisateur
   * @throws {ApiError} Si la transaction n'est pas trouvée (NOT_FOUND)
   *
   * @example
   * ```typescript
   * const status = await client.payment.getStatus('ORDER-001', 'CI')
   * if (status.status === 'SUCCESS') {
   *   // Paiement confirmé
   * }
   * ```
   */
  async getStatus(identifier, country) {
    const auth = this.resolveAuthenticator(country);
    try {
      const token = await auth.getToken();
      const raw = await this.httpClient.get(
        `/v1/payment/${encodeURIComponent(identifier)}`,
        token
      );
      return toPaymentStatus(raw);
    } catch (error) {
      if (error instanceof ApiError && this.isTokenExpired(error)) {
        const token = await auth.forceRefresh();
        const raw = await this.httpClient.get(
          `/v1/payment/${encodeURIComponent(identifier)}`,
          token
        );
        return toPaymentStatus(raw);
      }
      throw error;
    }
  }
  resolveAuthenticator(country) {
    const key = country.toUpperCase();
    const auth = this.authenticators.get(key);
    if (!auth) {
      throw new TypeError(
        `No credentials configured for country "${key}". Available: ${[...this.authenticators.keys()].join(", ")}`
      );
    }
    return auth;
  }
  isTokenExpired(error) {
    return error.apiCode === 1003 || error.apiStatus === "EXPIRED_TOKEN";
  }
};

// src/types/transfer.ts
function serializeTransferRequest(req) {
  return {
    currency: req.currency,
    merchant_transaction_id: req.merchantTransactionId,
    phone_number: req.phoneNumber,
    amount: req.amount,
    payment_method: req.paymentMethod,
    reason: req.reason,
    notify_url: req.notifyUrl
  };
}
function toTransferResponse(raw) {
  return {
    code: raw.code,
    status: raw.status,
    merchantTransactionId: raw.merchant_transaction_id,
    transactionId: raw.transaction_id,
    notifyToken: raw.notify_token,
    amount: String(raw.amount),
    feeAmount: String(raw.fee_amount),
    phoneNumber: raw.phone_number,
    currency: raw.currency
  };
}
function toTransferStatus(raw) {
  const user = raw.user;
  const result = {
    code: raw.code,
    status: raw.status,
    merchantTransactionId: raw.merchant_transaction_id,
    transactionId: raw.transaction_id,
    amount: String(raw.amount),
    feeAmount: String(raw.fee_amount)
  };
  if (user) {
    result.user = {
      name: user.name,
      email: user.email,
      phoneNumber: user.phone_number
    };
  }
  if (raw.phone_number) result.phoneNumber = raw.phone_number;
  if (raw.currency) result.currency = raw.currency;
  return result;
}

// src/api/transfer.ts
var TransferApi = class {
  constructor(httpClient, authenticators) {
    this.httpClient = httpClient;
    this.authenticators = authenticators;
  }
  /**
   * Effectue un transfert d'argent vers un numéro de téléphone.
   *
   * @param request - Paramètres du transfert (montant, devise, numéro, opérateur, etc.)
   * @param country - Code pays ISO (ex: 'CI', 'SN') — détermine les credentials utilisés
   * @returns Réponse contenant le statut initial et l'identifiant de transaction
   * @throws {ApiError} Si l'API retourne une erreur (ex: INSUFFICIENT_BALANCE, INVALID_PARAMS)
   * @throws {TypeError} Si le pays n'est pas configuré dans les credentials
   *
   * @example
   * ```typescript
   * const transfer = await client.transfer.create({
   *   currency: 'XOF',
   *   merchantTransactionId: 'TRANSFER-001',
   *   phoneNumber: '+2250707000001',
   *   amount: 500,
   *   paymentMethod: 'OM_CI',
   *   reason: 'Remboursement client',
   *   notifyUrl: 'https://monsite.com/webhook',
   * }, 'CI')
   * ```
   */
  async create(request, country) {
    validateTransferRequest(request);
    const auth = this.resolveAuthenticator(country);
    const body = serializeTransferRequest(request);
    try {
      const token = await auth.getToken();
      const raw = await this.httpClient.post(
        "/v1/transfer",
        body,
        token
      );
      return toTransferResponse(raw);
    } catch (error) {
      if (error instanceof ApiError && this.isTokenExpired(error)) {
        const token = await auth.forceRefresh();
        const raw = await this.httpClient.post(
          "/v1/transfer",
          body,
          token
        );
        return toTransferResponse(raw);
      }
      throw error;
    }
  }
  /**
   * Récupère le statut actuel d'un transfert.
   *
   * @param transactionId - Identifiant de la transaction CinetPay ou `merchant_transaction_id`
   * @param country - Code pays ISO (ex: 'CI', 'SN')
   * @returns Statut du transfert avec les informations du bénéficiaire
   * @throws {ApiError} Si la transaction n'est pas trouvée (NOT_FOUND)
   *
   * @example
   * ```typescript
   * const status = await client.transfer.getStatus('dc1f6d3d-...', 'CI')
   * if (status.status === 'SUCCESS') {
   *   console.log(`Transféré ${status.amount} ${status.currency}`)
   * }
   * ```
   */
  async getStatus(transactionId, country) {
    const auth = this.resolveAuthenticator(country);
    try {
      const token = await auth.getToken();
      const raw = await this.httpClient.get(
        `/v1/transfer/${encodeURIComponent(transactionId)}`,
        token
      );
      return toTransferStatus(raw);
    } catch (error) {
      if (error instanceof ApiError && this.isTokenExpired(error)) {
        const token = await auth.forceRefresh();
        const raw = await this.httpClient.get(
          `/v1/transfer/${encodeURIComponent(transactionId)}`,
          token
        );
        return toTransferStatus(raw);
      }
      throw error;
    }
  }
  resolveAuthenticator(country) {
    const key = country.toUpperCase();
    const auth = this.authenticators.get(key);
    if (!auth) {
      throw new TypeError(
        `No credentials configured for country "${key}". Available: ${[...this.authenticators.keys()].join(", ")}`
      );
    }
    return auth;
  }
  isTokenExpired(error) {
    return error.apiCode === 1003 || error.apiStatus === "EXPIRED_TOKEN";
  }
};

// src/types/balance.ts
function toBalance(raw) {
  return {
    code: raw.code,
    status: raw.status,
    availableBalance: raw.available_balance,
    currency: raw.currency
  };
}

// src/api/balance.ts
var BalanceApi = class {
  constructor(httpClient, authenticators) {
    this.httpClient = httpClient;
    this.authenticators = authenticators;
  }
  /**
   * Récupère le solde disponible du compte marchand pour un pays donné.
   * Ce solde est celui utilisé pour les opérations de transfert.
   *
   * @param country - Code pays ISO (ex: 'CI', 'SN') — détermine les credentials utilisés
   * @returns Solde disponible avec la devise du compte
   * @throws {ApiError} Si l'authentification échoue
   * @throws {TypeError} Si le pays n'est pas configuré dans les credentials
   *
   * @example
   * ```typescript
   * const balance = await client.balance.get('CI')
   * console.log(`${balance.availableBalance} ${balance.currency}`)
   * // "249711.74 XOF"
   * ```
   */
  async get(country) {
    const auth = this.resolveAuthenticator(country);
    try {
      const token = await auth.getToken();
      const raw = await this.httpClient.get(
        "/v1/balances",
        token
      );
      return toBalance(raw);
    } catch (error) {
      if (error instanceof ApiError && this.isTokenExpired(error)) {
        const token = await auth.forceRefresh();
        const raw = await this.httpClient.get(
          "/v1/balances",
          token
        );
        return toBalance(raw);
      }
      throw error;
    }
  }
  resolveAuthenticator(country) {
    const key = country.toUpperCase();
    const auth = this.authenticators.get(key);
    if (!auth) {
      throw new TypeError(
        `No credentials configured for country "${key}". Available: ${[...this.authenticators.keys()].join(", ")}`
      );
    }
    return auth;
  }
  isTokenExpired(error) {
    return error.apiCode === 1003 || error.apiStatus === "EXPIRED_TOKEN";
  }
};

// src/logger.ts
var ConsoleLogger = class {
  debug(message, data) {
    console.debug(`[cinetpay] ${message}`, data ?? "");
  }
  warn(message, data) {
    console.warn(`[cinetpay] ${message}`, data ?? "");
  }
  error(message, data) {
    console.error(`[cinetpay] ${message}`, data ?? "");
  }
};
var NoopLogger = class {
  debug() {
  }
  warn() {
  }
  error() {
  }
};

// src/client.ts
var CinetPayClient = class {
  /**
   * Crée un nouveau client CinetPay.
   *
   * @param config - Configuration avec les credentials par pays
   * @throws {TypeError} Si aucun credential n'est fourni
   * @throws {TypeError} Si le baseUrl n'utilise pas HTTPS (sauf localhost)
   */
  constructor(config) {
    const entries = Object.entries(config.credentials);
    if (entries.length === 0) {
      throw new TypeError(
        "At least one country credential must be provided in config.credentials"
      );
    }
    const tokenTtl = config.tokenTtl ?? DEFAULT_TOKEN_TTL;
    const timeout = config.timeout ?? DEFAULT_TIMEOUT;
    const tokenStore = config.tokenStore ?? new MemoryTokenStore();
    const logger = config.logger ?? (config.debug ? new ConsoleLogger() : new NoopLogger());
    const detectedEnv = this.detectEnvironment(entries, logger);
    const baseUrl = config.baseUrl ?? (detectedEnv === "live" ? PRODUCTION_BASE_URL : DEFAULT_BASE_URL);
    this.validateKeyUrlCoherence(detectedEnv, baseUrl, logger);
    if (!baseUrl.startsWith("https://") && !baseUrl.includes("localhost") && !baseUrl.includes("127.0.0.1")) {
      throw new TypeError(
        "baseUrl must use HTTPS for security. Use https:// or localhost for development."
      );
    }
    const ALLOWED_HOSTS = [
      "api.cinetpay.net",
      // Sandbox
      "api.cinetpay.co",
      // Production
      "localhost",
      "127.0.0.1"
    ];
    try {
      const url = new URL(baseUrl);
      if (!ALLOWED_HOSTS.some((h) => url.hostname === h || url.hostname.endsWith(`.${h}`))) {
        logger.warn(
          `baseUrl hostname "${url.hostname}" is not a known CinetPay domain. Expected: ${ALLOWED_HOSTS.join(", ")}. Proceeding anyway.`
        );
      }
    } catch {
      throw new TypeError(`baseUrl "${baseUrl}" is not a valid URL.`);
    }
    if (typeof globalThis.window !== "undefined" && typeof globalThis.document !== "undefined") {
      logger.warn(
        "Running in a browser environment. API credentials (api_key/api_password) will be exposed in client-side code. Use this SDK server-side only (Node.js, Deno, Bun) or via a backend proxy."
      );
    }
    if (config.forceIPv4) {
      this.applyIPv4(logger);
    }
    const httpClient = new HttpClient(baseUrl, timeout, logger);
    this.authenticators = /* @__PURE__ */ new Map();
    for (const [country, credentials] of entries) {
      const key = country.toUpperCase();
      this.authenticators.set(
        key,
        new Authenticator(httpClient, key, credentials, tokenTtl, tokenStore, logger)
      );
    }
    this.payment = new PaymentApi(httpClient, this.authenticators);
    this.transfer = new TransferApi(httpClient, this.authenticators);
    this.balance = new BalanceApi(httpClient, this.authenticators);
    logger.debug("CinetPayClient initialized", {
      countries: this.countries(),
      baseUrl,
      tokenTtl,
      timeout
    });
  }
  /**
   * Retourne la liste des codes pays configurés.
   *
   * @example
   * ```typescript
   * client.countries() // ['CI', 'SN']
   * ```
   */
  countries() {
    return [...this.authenticators.keys()];
  }
  /**
   * Révoque le token JWT en cache pour un pays donné.
   * Le prochain appel API déclenchera une nouvelle authentification.
   * Utile en cas de compromission ou de rotation manuelle des credentials.
   *
   * @param country - Code pays ISO (ex: 'CI', 'SN')
   * @throws {TypeError} Si le pays n'est pas configuré
   *
   * @example
   * ```typescript
   * await client.revokeToken('CI') // Force une ré-authentification au prochain appel
   * ```
   */
  async revokeToken(country) {
    const key = country.toUpperCase();
    const auth = this.authenticators.get(key);
    if (!auth) {
      throw new TypeError(
        `No credentials configured for country "${key}". Available: ${this.countries().join(", ")}`
      );
    }
    await auth.clearCache();
  }
  /**
   * Révoque tous les tokens JWT en cache pour tous les pays configurés.
   *
   * @example
   * ```typescript
   * await client.revokeAllTokens()
   * ```
   */
  async revokeAllTokens() {
    await Promise.all(
      [...this.authenticators.values()].map((auth) => auth.clearCache())
    );
  }
  /** Empêche la fuite des credentials via JSON.stringify */
  toJSON() {
    return { countries: this.countries() };
  }
  /**
   * Détecte l'environnement (test/live) à partir des préfixes des clés API.
   * - `sk_test_...` → 'test' (sandbox)
   * - `sk_live_...` → 'live' (production)
   * - Mixte → warning + 'mixed'
   * - Aucun préfixe reconnu → 'unknown'
   */
  detectEnvironment(entries, logger) {
    const envs = /* @__PURE__ */ new Set();
    for (const [country, creds] of entries) {
      if (creds.apiKey.startsWith(API_KEY_PREFIX_TEST)) {
        envs.add("test");
      } else if (creds.apiKey.startsWith(API_KEY_PREFIX_LIVE)) {
        envs.add("live");
      } else {
        logger.warn(
          `API key for country ${country.toUpperCase()} does not start with "${API_KEY_PREFIX_TEST}" or "${API_KEY_PREFIX_LIVE}". Expected format: sk_test_... (sandbox) or sk_live_... (production).`
        );
        envs.add("unknown");
      }
    }
    if (envs.has("test") && envs.has("live")) {
      logger.error(
        "MIXED ENVIRONMENTS: some credentials use sk_test_ (sandbox) and others use sk_live_ (production). This will cause authentication failures. Use the same environment for all countries."
      );
      return "mixed";
    }
    if (envs.has("live")) return "live";
    if (envs.has("test")) return "test";
    return "unknown";
  }
  /**
   * Vérifie la cohérence entre le type de clé API et le baseUrl.
   * - Clés `sk_test_` + baseUrl production → error
   * - Clés `sk_live_` + baseUrl sandbox → error
   */
  validateKeyUrlCoherence(detectedEnv, baseUrl, logger) {
    const isSandboxUrl = baseUrl.includes("cinetpay.net");
    const isProductionUrl = baseUrl.includes("cinetpay.co");
    if (detectedEnv === "live" && isSandboxUrl) {
      logger.error(
        `ENVIRONMENT MISMATCH: production keys (sk_live_) are used with sandbox URL (${baseUrl}). Use baseUrl: "${PRODUCTION_BASE_URL}" for production, or remove baseUrl to auto-detect.`
      );
    }
    if (detectedEnv === "test" && isProductionUrl) {
      logger.error(
        `ENVIRONMENT MISMATCH: sandbox keys (sk_test_) are used with production URL (${baseUrl}). Use baseUrl: "${DEFAULT_BASE_URL}" for sandbox, or remove baseUrl to auto-detect.`
      );
    }
  }
  /**
   * Force la résolution DNS en IPv4 dans Node.js.
   * Utilise `dns.setDefaultResultOrder('ipv4first')` — disponible depuis Node.js 16.4.
   * No-op silencieux dans les environnements non-Node (navigateurs, Deno, Bun).
   */
  applyIPv4(logger) {
    try {
      const dns = __require("dns");
      if (typeof dns.setDefaultResultOrder === "function") {
        dns.setDefaultResultOrder("ipv4first");
        logger.debug("DNS resolution forced to IPv4");
      } else {
        logger.warn("forceIPv4: dns.setDefaultResultOrder not available in this Node.js version");
      }
    } catch {
      logger.debug("forceIPv4: not in Node.js, option ignored");
    }
  }
};

// src/constants/statuses.ts
var TRANSACTION_STATUSES = {
  /** Opération réussie (réponse API, pas un statut final de transaction) */
  OK: "OK",
  /** Transaction traitée avec succès — statut final */
  SUCCESS: "SUCCESS",
  /** Une erreur s'est produite */
  OPERATION_ERROR: "OPERATION_ERROR",
  /** Ressource introuvable */
  NOT_FOUND: "NOT_FOUND",
  /** Identifiants API invalides */
  INVALID_CREDENTIALS: "INVALID_CREDENTIALS",
  /** Paramètres de requête invalides */
  INVALID_PARAMS: "INVALID_PARAMS",
  /** Token JWT expiré (le SDK gère le renouvellement automatiquement) */
  EXPIRED_TOKEN: "EXPIRED_TOKEN",
  /** Token JWT invalide */
  INVALID_TOKEN: "INVALID_TOKEN",
  /** La transaction existe déjà (merchant_transaction_id dupliqué) — statut final */
  TRANSACTION_EXIST: "TRANSACTION_EXIST",
  /** Transaction initiée, en attente d'action de l'utilisateur */
  INITIATED: "INITIATED",
  /** L'utilisateur a commencé le processus de paiement */
  PENDING: "PENDING",
  /** Transaction expirée (l'utilisateur n'a pas confirmé) */
  EXPIRED: "EXPIRED",
  /** Code OTP incorrect */
  OTP_ERROR: "OTP_ERROR",
  /** Code OTP expiré */
  OTP_EXPIRED: "OTP_EXPIRED",
  /** Solde marchand insuffisant — statut final */
  INSUFFICIENT_BALANCE: "INSUFFICIENT_BALANCE",
  /** Utilisateur introuvable */
  USER_NOT_FOUND: "USER_NOT_FOUND",
  /** Utilisateur bloqué */
  USER_IS_BLOCKED: "USER_IS_BLOCKED",
  /** Paiement échoué — statut final */
  FAILED: "FAILED",
  /** IP non autorisée (whitelist) */
  NOT_ALLOWED: "NOT_ALLOWED"
};
var FINAL_STATUSES = /* @__PURE__ */ new Set([
  "SUCCESS",
  "FAILED",
  "TRANSACTION_EXIST",
  "INSUFFICIENT_BALANCE"
]);
function isFinalStatus(status) {
  return FINAL_STATUSES.has(status);
}
var API_CODES = {
  200: "OK",
  100: "SUCCESS",
  [-1]: "OPERATION_ERROR",
  404: "NOT_FOUND",
  1005: "INVALID_CREDENTIALS",
  1004: "INVALID_PARAMS",
  1003: "EXPIRED_TOKEN",
  1002: "INVALID_TOKEN",
  1200: "TRANSACTION_EXIST",
  2001: "INITIATED",
  2002: "PENDING",
  2003: "EXPIRED",
  2004: "OTP_ERROR",
  2008: "OTP_EXPIRED",
  2005: "INSUFFICIENT_BALANCE",
  2006: "USER_NOT_FOUND",
  2007: "USER_IS_BLOCKED",
  2010: "FAILED",
  2011: "NOT_ALLOWED"
};
var COUNTRY_CODES = [
  "CI",
  "BF",
  "ML",
  "SN",
  "TG",
  "GN",
  "CM",
  "BJ",
  "CD",
  "NE"
];

// src/types/webhook.ts
function toWebhookPayload(raw) {
  const user = raw.user;
  return {
    notifyToken: raw.notify_token,
    merchantTransactionId: raw.merchant_transaction_id,
    transactionId: raw.transaction_id,
    ...user && {
      user: {
        name: user.name,
        email: user.email,
        phoneNumber: user.phone_number
      }
    }
  };
}

// src/webhooks/notification.ts
function verifyNotification(expectedToken, receivedToken) {
  if (expectedToken.length !== receivedToken.length) return false;
  let result = 0;
  for (let i = 0; i < expectedToken.length; i++) {
    result |= expectedToken.charCodeAt(i) ^ receivedToken.charCodeAt(i);
  }
  return result === 0;
}
function parseNotification(body) {
  const data = typeof body === "string" ? JSON.parse(body) : body;
  if (!data.notify_token || !data.transaction_id) {
    throw new TypeError(
      "Invalid webhook payload: missing notify_token or transaction_id"
    );
  }
  return toWebhookPayload(data);
}

export { API_CODES, API_KEY_PREFIX_LIVE, API_KEY_PREFIX_TEST, ApiError, AuthenticationError, CHANNELS, COUNTRY_CODES, CURRENCIES, CinetPayClient, CinetPayError, ConsoleLogger, DEFAULT_BASE_URL, DEFAULT_TIMEOUT, DEFAULT_TOKEN_TTL, MemoryTokenStore, NetworkError, NoopLogger, PAYMENT_METHODS, PAYMENT_METHODS_BY_COUNTRY, PRODUCTION_BASE_URL, TRANSACTION_STATUSES, TimeoutError, ValidationError, isFinalStatus, parseNotification, verifyNotification };
